active/0000-extensible-commands.md

```diff
@@ -0,0 +1,64 @@
+1. `mocha` provides a `--grep/-g` option to run tests that match a supplied regex. This is different behavior from the standard `--filter` option allowed by `ember test`. Right now, the `--grep` feature is essentially unusable without custom test commands (i.e. `ember mocha:test`) or hacking the `--filter` option (i.e. `--filter="^"` indicates a regex). Ideally, `ember-cli-mocha` would be able to extend the built-in `test` command to add a `--grep` option.
+2. An addon with a post-build hook that wants to modify the build output (e.g. feature flags) might want to accept additional options.
+
+In generally, these use cases are examples of the same idea: the Ember CLI built-in commands perform common tasks that might need different implementations across projects.
+
+# Detailed design
+
+Addons can currently specify commands to extend the CLI interface via the `includedCommands` hook. When a command is invoked, the `lib/cli/lookup-command.js` searches through all the loaded addon commands, as well as built-in commands, for a match.
+
+This lookup process would be refactored to use the DAG structure of loaded addons to build a flattened list of addons that implement the requested command. This stack would then be recursively invoked, such that each addon's version of the command can call `this._super()` to pass control to the next addon in the "stack". The bottom of the stack would be the built-in implementation (if one exists).
+
+This structure is analagous to Connect-style middleware, where each function in the stack can decide to pass control further, or halt the recursion by not calling `this._super()`.
+
+Here's a rough sketch in code:
+
+```
+var command, commandOptions;
+var orderedAddonList = project.addons;
+var commandStack = orderedAddonList.filter(function(addon) {
+        return addon.hasCommand(command);
+    }).map(function(addon) {
+        return addon.getCommand(command);
+    });
+
+var context = { _super: invokeNextCommandInStack };
+function invokeNextCommandInStack() {
+    var nextCommand = commandStack.shift();
+    return nextCommand.call(context, command, commandOptions);
+}
+
+invokeNextCommandInStack()
+```
+
+> **Note:** obviously, there's a lot more that should happen, and the commands are not functions themselves. Likely, either the Command model would need to incorporate the idea of a "stack", the `/lib/cli/cli.js` file would have to become stack-aware, or we introduce a new CommandStack abstraction to handle this.
+
+We would also add a `-v/--verbose` flag which would print out information about what extensions are running. Output could show what addon in the "stack" is currently executing, which would make it easier to track down what code could be introducing a bug.
+
+# Drawbacks
+
+* Extending built-in commands could result in user confusion, i.e. I checkout an Ember CLI project and run `ember test` and it performs unexpected behaviors, so I Google around for "ember test" and none of the standard docs help me.
```

Comment:
but I wonder if it should just be 100% replacement, not inheritance merge. Inheritance here is tricky, i suspect it will just get sloppy.